src/shims/sync.rs

```diff
@@ -0,0 +1,519 @@
+        } else {
+            mutexattr_get_kind(this, attr_op)?.not_undef()?
+        };
+
+        mutex_set_locked_count(this, mutex_op, Scalar::from_u32(0))?;
+        mutex_set_kind(this, mutex_op, kind)?;
+
+        Ok(0)
+    }
+
+    fn pthread_mutex_lock(&mut self, mutex_op: OpTy<'tcx, Tag>) -> InterpResult<'tcx, i32> {
+        let this = self.eval_context_mut();
+
+        let mutex = this.read_scalar(mutex_op)?.not_undef()?;
+        if this.is_null(mutex)? {
+            return this.eval_libc_i32("EINVAL");
+        }
+
+        let kind = mutex_get_kind(this, mutex_op)?.not_undef()?;
+        let locked_count = mutex_get_locked_count(this, mutex_op)?.to_u32()?;
+
+        if kind == this.eval_libc("PTHREAD_MUTEX_NORMAL")? {
+            if locked_count == 0 {
+                mutex_set_locked_count(this, mutex_op, Scalar::from_u32(1))?;
+                Ok(0)
+            } else {
+                throw_machine_stop!(TerminationInfo::Deadlock);
+            }
+        } else if kind == this.eval_libc("PTHREAD_MUTEX_ERRORCHECK")? {
+            if locked_count == 0 {
+                mutex_set_locked_count(this, mutex_op, Scalar::from_u32(1))?;
+                Ok(0)
+            } else {
+                this.eval_libc_i32("EDEADLK")
+            }
+        } else if kind == this.eval_libc("PTHREAD_MUTEX_RECURSIVE")? {
+            match locked_count.checked_add(1) {
+                Some(new_count) => {
+                    mutex_set_locked_count(this, mutex_op, Scalar::from_u32(new_count))?;
+                    Ok(0)
+                }
+                None => this.eval_libc_i32("EAGAIN"),
+            }
+        } else {
+            this.eval_libc_i32("EINVAL")
+        }
+    }
+
+    fn pthread_mutex_trylock(&mut self, mutex_op: OpTy<'tcx, Tag>) -> InterpResult<'tcx, i32> {
+        let this = self.eval_context_mut();
+
+        let mutex = this.read_scalar(mutex_op)?.not_undef()?;
+        if this.is_null(mutex)? {
+            return this.eval_libc_i32("EINVAL");
+        }
+
+        let kind = mutex_get_kind(this, mutex_op)?.not_undef()?;
+        let locked_count = mutex_get_locked_count(this, mutex_op)?.to_u32()?;
+
+        if kind == this.eval_libc("PTHREAD_MUTEX_NORMAL")?
+            || kind == this.eval_libc("PTHREAD_MUTEX_ERRORCHECK")?
+        {
+            if locked_count == 0 {
+                mutex_set_locked_count(this, mutex_op, Scalar::from_u32(1))?;
+                Ok(0)
+            } else {
+                this.eval_libc_i32("EBUSY")
+            }
+        } else if kind == this.eval_libc("PTHREAD_MUTEX_RECURSIVE")? {
+            match locked_count.checked_add(1) {
+                Some(new_count) => {
+                    mutex_set_locked_count(this, mutex_op, Scalar::from_u32(new_count))?;
+                    Ok(0)
+                }
+                None => this.eval_libc_i32("EAGAIN"),
+            }
+        } else {
+            this.eval_libc_i32("EINVAL")
```

Comment:
I feel like this also should raise UB -- clearly the client is breaking protocol if we end up with another mutex kind here. (Same for lock, unlock.)